src/main.c

```diff
@@ -2042,6 +2112,43 @@ static lbm_value ext_set_fw_version(lbm_value *args, lbm_uint argn) {
     return VESC_IF->lbm_enc_sym_true;
 }
 
+// Called from Lisp to pass in values from the bms. If no prameters are called, return
+// tiltback_bms_enabled.
+static lbm_value ext_bms(lbm_value *args, lbm_uint argn) {
+    Data *d = (Data *) ARG;
+    d->bms_fault = BMSF_NONE;
+
+    if (argn == 0 || !d->float_conf.tiltback_bms_enabled) {
+        return d->float_conf.tiltback_bms_enabled;
+    }
+
+    if (VESC_IF->lbm_dec_as_float(args[5]) > d->float_conf.tiltback_bms_msg) {
+        bms_set_fault(&d->bms_fault, BMSF_CONNECTION);
+        return d->float_conf.tiltback_bms_enabled;
+    }
+    if (VESC_IF->lbm_dec_as_float(args[0]) < d->float_conf.tiltback_cell_lv) {
+        bms_set_fault(&d->bms_fault, BMSF_CELL_UNDER_VOLTAGE);
+    }
+    if (VESC_IF->lbm_dec_as_float(args[1]) > d->float_conf.tiltback_cell_hv) {
+        bms_set_fault(&d->bms_fault, BMSF_CELL_OVER_VOLTAGE);
+    }
+    if (VESC_IF->lbm_dec_as_i32(args[2]) < d->float_conf.tiltback_cell_lt) {
+        bms_set_fault(&d->bms_fault, BMSF_CELL_UNDER_TEMP);
+    }
+    if (VESC_IF->lbm_dec_as_i32(args[3]) > d->float_conf.tiltback_cell_ht) {
+        bms_set_fault(&d->bms_fault, BMSF_CELL_OVER_TEMP);
+    }
+    if (VESC_IF->lbm_dec_as_i32(args[4]) > d->float_conf.tiltback_bms_ht) {
+        bms_set_fault(&d->bms_fault, BMSF_OVER_TEMP);
+    }
+    if (fabsf(VESC_IF->lbm_dec_as_float(args[0]) - VESC_IF->lbm_dec_as_float(args[1])) >
+        d->float_conf.tiltback_cell_bal) {
+        bms_set_fault(&d->bms_fault, BMSF_CELL_BALANCE);
+    }
```

Comment:
I understand putting the logic here is the most straightforward, but there should be a separation between the passing of the data from lisp and the evaluation logic. What seems the best approach to me is declaring a struct with all the values passed in from lisp. In this function, just fill the struct with values and then have a function like `bms_update_fault(uint32_t *fault_mask, TheNewStruct *bms_data);`